agent/xds/listeners.go

```diff
@@ -176,39 +185,141 @@ func makePublicListener(cfgSnap *proxycfg.ConfigSnapshot, token string) (proto.M
+		// default config if there is an error so it's safe to continue.
+		s.Logger.Printf("[WARN] envoy: failed to parse Upstream[%s].Config: %s",
+			u.Identifier(), err)
+	}
+	if cfg.ListenerJSON != "" {
+		return makeListenerFromUserConfig(cfg.ListenerJSON)
 	}
+
 	addr := u.LocalBindAddress
 	if addr == "" {
 		addr = "127.0.0.1"
 	}
+
 	l := makeListener(u.Identifier(), addr, u.LocalBindPort)
-	tcpProxy, err := makeTCPProxyFilter(u.Identifier(), u.Identifier())
+	filter, err := makeListenerFilter(cfg.Protocol, u.Identifier(), u.Identifier(), "upstream_", false)
 	if err != nil {
-		return l, err
+		return nil, err
 	}
 	l.FilterChains = []envoylistener.FilterChain{
 		{
 			Filters: []envoylistener.Filter{
-				tcpProxy,
+				filter,
 			},
 		},
 	}
 	return l, nil
 }
 
-func makeTCPProxyFilter(name, cluster string) (envoylistener.Filter, error) {
+func makeListenerFilter(protocol, filterName, cluster, statPrefix string, ingress bool) (envoylistener.Filter, error) {
+	switch protocol {
+	case "grpc":
+		return makeHTTPFilter(filterName, cluster, statPrefix, ingress, true, true)
+	case "http2":
+		return makeHTTPFilter(filterName, cluster, statPrefix, ingress, false, true)
+	case "http":
+		return makeHTTPFilter(filterName, cluster, statPrefix, ingress, false, false)
+	case "tcp":
+		fallthrough
+	default:
+		return makeTCPProxyFilter(filterName, cluster, statPrefix)
+	}
+}
+
+func makeTCPProxyFilter(filterName, cluster, statPrefix string) (envoylistener.Filter, error) {
 	cfg := &envoytcp.TcpProxy{
-		StatPrefix: name,
+		StatPrefix: makeStatPrefix("tcp", statPrefix, filterName),
 		Cluster:    cluster,
 	}
 	return makeFilter("envoy.tcp_proxy", cfg)
 }
 
+func makeStatPrefix(protocol, prefix, filterName string) string {
+	// Replace colons here because Envoy does that in the metrics for the actual
+	// clusters but doesn't in the stat prefix here while dashboards assume they
+	// will match.
+	return fmt.Sprintf("%s%s_%s", prefix, strings.Replace(filterName, ":", "_", -1), protocol)
+}
+
+func makeHTTPFilter(filterName, cluster, statPrefix string, ingress, grpc, http2 bool) (envoylistener.Filter, error) {
+	op := envoyhttp.INGRESS
+	if !ingress {
+		op = envoyhttp.EGRESS
+	}
+	proto := "http"
+	if grpc {
+		proto = "grpc"
+	}
+	cfg := &envoyhttp.HttpConnectionManager{
+		StatPrefix: makeStatPrefix(proto, statPrefix, filterName),
+		CodecType:  envoyhttp.AUTO,
+		RouteSpecifier: &envoyhttp.HttpConnectionManager_RouteConfig{
+			RouteConfig: &envoy.RouteConfiguration{
+				Name: filterName,
+				VirtualHosts: []envoyroute.VirtualHost{
+					envoyroute.VirtualHost{
+						Name:    filterName,
+						Domains: []string{"*"},
+						Routes: []envoyroute.Route{
+							envoyroute.Route{
+								Match: envoyroute.RouteMatch{
+									PathSpecifier: &envoyroute.RouteMatch_Prefix{
+										Prefix: "/",
+									},
+									// TODO(banks) Envoy supports matching only valid GRPC
```

Comment:
Is this TODO still accurate? Didn't you validate newer envoy versions recently and do we now only support versions new enough for this?

Reply:
Yes. Our version of the Envoy protobuf API (SDK here in comment) doesn't have the fields we need to add this feature right now although all our supported Envoy versions to support it supposedly.

We support 1.8.0+ now but I didn't want to re-vendor a later but old-enough version and re-test everything just to pull that in since I know master is incompatible and we need to fix that right after this release anyway.

This is a pretty simple feature to add back and not really a big deal in the interim.